server/api/router/auth.js

```diff
@@ -1,20 +1,71 @@
-const router = require('express').Router();
-const { User } = require('../../db/index');
-
-router.post('/login', (req, res, next) => {
-  const { email, password } = req.body;
-  User.findAll({ where: { email: email, password: password } })
-    .then(user => {
-      if (user) {
-        req.session.userId = user.id;
-        res.json(user);
-      } else {
-        const error = new Error('Wrong password or email!');
-        error.status = 401;
-        next(err);
-      }
-    })
-    .catch(next);
-});
-
-module.exports = router;
+const router = require('express').Router();
+const { User } = require('../../db/index');
+
+router.post('/login', async (req, res) => {
+  try {
+      const { email, password } = req.body;
+      if (email && password) {
+          const user = await User.findOne({ where: { email, password } });
+              if (!user) {
+                  res.sendStatus(401);
+                  console.log('Log in failed; invalid credentials');
+              } else if (user.loggedIn) {
+                res.sendStatus(200);
+                console.log('Already logged in as:', user.id);
+                } else {
+                const loggedInUser = await user.update({loggedIn: true});
+                res.json(loggedInUser);
+                console.log('Logged in sucessfully as:', user.id);
+              }
+      } else {
+        res.sendStatus(500)
+        console.log('Email and password required');
+      }
+  } catch (e) {
+      res.sendStatus(500)
+      console.log('Error logging in', e);
+  }
+});
+
+router.post('/signup', async (req, res) => {
+  try {
+      const guestUser = await User.findOne({where: {id: req.userId}});
+      const { firstName, lastName, email, password, addressLine1, addressLine2, city, state, zipcode, phone } = req.body;
+
+      if (firstName && lastName && email && password) {
+          const user = await User.findOne({ where: { email } });
+              if (user) {
+                  res.sendStatus(401);
+                  console.log('Email already exists in the system');
+              } else {
+                  const registeredUser = await guestUser.update({type: 'registered', loggedIn: true, firstName, lastName, email, password, addressLine1, addressLine2, city, state, zipcode, phone});
+                  res.json(registeredUser);
+                  console.log('Signed up and logged in as:', registeredUser.id);
+              }
+      } else {
+        res.sendStatus(500);
+        console.log('Required fields missing for signup');
+      }
+  } catch (e) {
+      console.log('Error logging in', e);
+  }
+});
+
+router.post('/logout', async (req, res) => {
+  try {
+  const user = await User.findOne({where: {id: req.userId}});
+    if (user) {
+      const loggedOutUser = await user.update({loggedIn: false});
```

Comment:
You don't need to do any of this! All you need to do is destroy the session.

```javascript
// Clear the session from the database.
const session = await Session.findByPk(req.session.id);
await session.destroy();
// Clear the session from memory...
req.session.destroy(err => {
  if (err) {
    res.sendStatus(500);
  } else {
    // Its worth note you haven't yet put this on process.env, so this will have to be manual for now.
    // Even though we've clear the session, on the next request it would see the same cookie again, thus, recreating the session with the same SID again - thats not good at all. So lets dump the cookie too.
    res.clearCookie(process.env.SESH_NAME);
    // Lets send em home in case they were already sitting somewhere that required authentication.
    res.redirect('/');
  }
});
```